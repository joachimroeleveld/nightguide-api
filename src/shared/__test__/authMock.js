const { TEST_USER_1 } = require('../../shared/__test__/fixtures/userFixtures');
const auth = require('../../framework/auth');
const User = require('../../routes/users/userModel');

function mockUserAuth(sandbox = global.sandbox) {
  const user = new User(TEST_USER_1);
  user.checkRole = sandbox.stub().returns(true);
  sandbox.stub(auth, 'getUserFromToken').resolves(user);
}

function restoreUserAuth() {
  auth.getUserFromToken.restore();
}

function mockAppClientAuth(sandbox = global.sandbox) {
  sandbox.stub(auth, 'checkAppTokenHeader').returns(true);
}

function restoreAppClientAuth() {
  auth.checkAppTokenHeader.restore();
}

module.exports = {
  mockUserAuth,
  restoreUserAuth,
  mockAppClientAuth,
  restoreAppClientAuth,
};
