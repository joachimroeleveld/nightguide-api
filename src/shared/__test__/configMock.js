const config = require('../config');

const mockedEnv = {
  ...process.env,
};

function mockKey(key, value, sandbox = global.sandbox) {
  mockedEnv[key] = value;
  sandbox.stub(config, 'get').callsFake(key => mockedEnv[key]);
}

module.exports = {
  mockKey,
};
