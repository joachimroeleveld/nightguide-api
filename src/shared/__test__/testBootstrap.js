const dotenv = require('dotenv');
dotenv.load();

const sgMail = require('@sendgrid/mail');
const mongoose = require('mongoose');
const config = require('../config');

const sandbox = require('sinon').createSandbox();
const { mockAuth } = require('./authMock');
const configMock = require('./configMock');

const { createExpressApp } = require('../../framework/expressServer');

beforeAll(cb => {
  // Use global sandbox for mocks that you want to persist across multiple test suites
  if (!global.sandbox) {
    global.sandbox = sandbox;
  }

  mockAuth();

  configMock.mockKey('HOST', 'http://localhost:8080');

  // Mock SendGrid
  sandbox.stub(sgMail, 'send').resolves();

  global.app = createExpressApp();

  mongoose
    .connect(
      config.get('MONGO_URI'),
      { useNewUrlParser: true, dbName: 'main-test' }
    )
    .then(() => cb())
    .catch(err => cb(err));
});

afterAll(async () => {
  mongoose.connection.close();

  global.sandbox.restore();
});
