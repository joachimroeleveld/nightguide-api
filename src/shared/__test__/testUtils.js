const mongoose = require('mongoose');
const _ = require('lodash');

const { deepFreeze } = require('../util/objects');

function generateMongoFixture(base, overrides) {
  return deepFreeze({
    ...base,
    _id: new mongoose.Types.ObjectId().toString(),
    ...overrides,
  });
}

async function resetDb() {
  const removals = Object.keys(mongoose.connection.collections).map(
    collection =>
      new Promise(resolve =>
        mongoose.connection.collections[collection].remove(resolve)
      )
  );

  let dropIndices = [];
  // dropIndices = Object.keys(mongoose.connection.collections).map(
  //   collection =>
  //     new Promise(resolve =>
  //       mongoose.connection.collections[collection].dropAllIndexes(resolve)
  //     )
  // );

  await Promise.all(_.flatten([removals, dropIndices]));
}

function setFixtureLocation(fixture, coordinates) {
  return {
    ...fixture,
    location: {
      ...fixture.location,
      coordinates: {
        type: 'Point',
        coordinates: coordinates,
      },
    },
  };
}

module.exports = {
  resetDb,
  setFixtureLocation,
  generateMongoFixture,
};
