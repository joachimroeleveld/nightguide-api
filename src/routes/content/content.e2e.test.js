require('../../shared/__test__/testBootstrap');

const request = require('supertest');
const sinon = require('sinon');

const { validator } = require('../../shared/openapi');
const {
  TEST_CONTENT_1,
  TEST_CONTENT_2,
  generateMongoFixture,
} = require('../../shared/__test__/fixtures');
const { CONTENT_TYPES } = require('../../shared/constants');
const { resetDb } = require('../../shared/__test__/testUtils');
const contentRepository = require('./contentRepository');

const CONTENT_SNAPSHOT_MATCHER = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
};

const sandbox = sinon.createSandbox();

describe('content e2e', () => {
  beforeEach(async () => {
    sandbox.restore();
    await resetDb();
  });

  describe('GET /content/:contentType', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/content/{contentType}'
    );

    it('happy path', async () => {
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_1
      );

      const res = await request(global.app).get(
        `/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`
      );

      expect(res.status).toEqual(200);
      expect(res.body.totalCount).toEqual(1);
      expect(res.body.results[0]).toMatchInlineSnapshot(
        CONTENT_SNAPSHOT_MATCHER,
        `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "pageSlug": "nl/utrecht",
  "title": Object {
    "en": "Pretty interesting article",
  },
  "updatedAt": Any<String>,
  "urlSlugs": Array [
    "pretty-interesting-article",
  ],
}
`
      );
      expect(validateResponse(res)).toBeUndefined();
    });

    it('should limit the amount of results to the limit parameter', async () => {
      const content1 = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_1
      );
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_2
      );

      const res = await request(global.app)
        .get(`/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`)
        .query({
          limit: 1,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toEqual(1);
      expect(res.body.limit).toEqual(1);
      expect(res.body.results[0].id).toEqual(content1._id.toString());
      expect(validateResponse(res)).toBeUndefined();
    });

    it('should skip items set in offset parameter', async () => {
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_1
      );
      const content2 = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_2
      );

      const res = await request(global.app)
        .get(`/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`)
        .query({
          offset: 1,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toEqual(1);
      expect(res.body.offset).toEqual(1);
      expect(res.body.results[0].id).toEqual(content2._id.toString());
      expect(validateResponse(res)).toBeUndefined();
    });

    it('returns a totalcount with each result', async () => {
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_1
      );
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_2
      );

      const res = await request(global.app)
        .get(`/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`)
        .query({ limit: 1 });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(1);
      expect(res.body.totalCount).toBe(2);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('ids filter', async () => {
      const content1 = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_1
      );
      const content2 = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_2
      );
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        generateMongoFixture(TEST_CONTENT_1)
      );

      const ids = [content1._id.toString(), content2._id.toString()].sort();

      const res = await request(global.app)
        .get(`/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`)
        .query({
          ids,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(2);
      expect(res.body.results.map(item => item.id).sort()).toEqual(ids);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('pageSlug filter', async () => {
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        {
          ...TEST_CONTENT_1,
          pageSlug: 'nl/utrecht',
        }
      );
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        {
          ...TEST_CONTENT_2,
          pageSlug: 'es/ibiza',
        }
      );

      const res = await request(global.app)
        .get(`/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`)
        .query({
          pageSlug: 'nl/utrecht',
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(1);
      expect(res.body.results[0].name).toBe(TEST_CONTENT_1.name);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('GET /content/:contentType/:contentId', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/content/{contentType}/{contentId}'
    );

    it('happy path', async () => {
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_1
      );

      const res = await request(global.app)
        .get(
          `/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}/${
            TEST_CONTENT_1._id
          }`
        )
        .send()
        .expect(200);

      expect(res.body.id).toEqual(TEST_CONTENT_1._id);
      expect(res.body).toMatchSnapshot(CONTENT_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('GET /content/:contentType/slug/:slug', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/content/{contentType}/slug/{slug}'
    );

    it('happy path', async () => {
      const content1 = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        {
          ...TEST_CONTENT_1,
          urlSlugs: ['test'],
        }
      );

      const res = await request(global.app)
        .get(
          `/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}/slug/${
            content1.urlSlug
          }`
        )
        .send()
        .expect(200);

      expect(res.body.urlSlugs[0]).toEqual(content1.urlSlug);
      expect(res.body.id).toEqual(content1._id.toString());
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('POST /content/:contentType', () => {
    const validateResponse = validator.validateResponse(
      'post',
      '/content/{contentType}'
    );

    it('happy path', async () => {
      const res = await request(global.app)
        .post(`/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`)
        .send(contentRepository.deserialize(TEST_CONTENT_1));
      const body = res.body;

      expect(res.status).toEqual(201);
      expect(body.id).toEqual(TEST_CONTENT_1._id);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('generates a URL slug from the title if not provided', async () => {
      const res = await request(global.app)
        .post(`/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`)
        .send({
          ...contentRepository.deserialize(TEST_CONTENT_1),
          title: { en: ' This is a test ' },
        });

      expect(res.status).toEqual(201);
      expect(res.body.urlSlugs[0]).toEqual('this-is-a-test');
      expect(validateResponse(res)).toBeUndefined();
    });

    it("doesn't overwrite the URL slug if provided", async () => {
      const res = await request(global.app)
        .post(`/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}`)
        .send({
          ...contentRepository.deserialize(TEST_CONTENT_1),
          urlSlugs: ['test'],
        });

      expect(res.status).toEqual(201);
      expect(res.body.urlSlugs[0]).toEqual('test');
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('PUT /content/:contentType', () => {
    const validateResponse = validator.validateResponse(
      'put',
      '/content/{contentType}/{contentId}'
    );

    it('happy path', async () => {
      const content1 = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_1
      );

      const res = await request(global.app)
        .put(
          `/content/${CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE}/${
            content1._id
          }`
        )
        .send({
          ...contentRepository.deserialize(TEST_CONTENT_1),
          title: {
            en: 'changed',
          },
        });
      const body = res.body;

      expect(res.status).toEqual(200);
      expect(body.id).toEqual(TEST_CONTENT_1._id);
      expect(body.title.en).toEqual('changed');
      expect(validateResponse(res)).toBeUndefined();
    });

    it('ensures a unique URL slug', async () => {
      await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        {
          ...TEST_CONTENT_1,
          urlSlugs: ['test'],
        }
      );
      const content2 = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_2
      );

      const res = await request(global.app)
        .put(
          `/content/${
            CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE
          }/${content2._id.toString()}`
        )
        .send({
          ...contentRepository.deserialize(TEST_CONTENT_2),
          urlSlugs: ['test'],
        });

      const body = res.body;
      expect(res.status).toEqual(200);
      expect(body.id).toEqual(content2._id.toString());
      expect(body.urlSlugs[0]).toEqual('test-2');
      expect(validateResponse(res)).toBeUndefined();
    });

    it('preserves the old slug when if changed', async () => {
      const content1 = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        {
          ...TEST_CONTENT_1,
          urlSlugs: ['foo'],
        }
      );

      const res = await request(global.app)
        .put(
          `/content/${
            CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE
          }/${content1._id.toString()}`
        )
        .send({
          ...contentRepository.deserialize(TEST_CONTENT_1),
          urlSlugs: ['bar'],
        });

      const body = res.body;
      expect(res.status).toEqual(200);
      expect(body.id).toEqual(content1._id.toString());
      expect(body.urlSlugs[0]).toEqual('bar');
      expect(body.urlSlugs[1]).toEqual('foo');
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('DELETE /content/:contentType/:contentId', () => {
    const validateResponse = validator.validateResponse(
      'delete',
      '/content/{contentType}/{contentId}'
    );

    it('happy path', async () => {
      const createdTag = await contentRepository.createContent(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        TEST_CONTENT_1
      );

      const res = await request(global.app).delete(
        `/content/${
          CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE
        }/${createdTag._id.toString()}`
      );

      let content = await contentRepository.getContentSingle(
        CONTENT_TYPES.CONTENT_TYPE_VENUES_ARTICLE,
        createdTag._id.toString()
      );

      expect(content).toBe(null);
      expect(res.status).toEqual(200);
      expect(res.body.success).toEqual(true);
      expect(validateResponse(res)).toBeUndefined();
    });
  });
});
