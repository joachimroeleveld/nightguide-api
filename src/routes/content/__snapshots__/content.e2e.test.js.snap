// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content e2e GET /content/:contentType/:contentId happy path 1`] = `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "pageSlug": "nl/utrecht",
  "title": Object {
    "en": "Pretty interesting article",
  },
  "updatedAt": Any<String>,
  "urlSlugs": Array [
    "pretty-interesting-article",
  ],
}
`;
